codex/stores/repostore/operations.nim

```diff
@@ -0,0 +1,213 @@
+
+      self.totalBlocks = count
+      codex_repostore_blocks.set(count.int64)
+      count.some
+  )
+
+proc updateQuotaUsage*(
+  self: RepoStore,
+  plusUsed: NBytes = 0.NBytes,
+  minusUsed: NBytes = 0.NBytes,
+  plusReserved: NBytes = 0.NBytes,
+  minusReserved: NBytes = 0.NBytes
+): Future[?!void] {.async.} =
+  await self.metaDs.modify(QuotaUsedKey,
+    proc (maybeCurrUsage: ?QuotaUsage): Future[?QuotaUsage] {.async.} =
+      var usage: QuotaUsage
+
+      if currUsage =? maybeCurrUsage:
+        usage = QuotaUsage(used: currUsage.used + plusUsed - minusUsed, reserved: currUsage.reserved + plusReserved - minusReserved)
+      else:
+        usage = QuotaUsage(used: plusUsed - minusUsed, reserved: plusReserved - minusReserved)
+
+      if usage.used + usage.reserved > self.quotaMaxBytes:
+        raise newException(QuotaNotEnoughError,
+          "Quota usage would exceed the limit. Used: " & $usage.used & ", reserved: " &
+            $usage.reserved & ", limit: " & $self.quotaMaxBytes)
+      else:
+        self.quotaUsage = usage
+        codex_repostore_bytes_used.set(usage.used.int64)
+        codex_repostore_bytes_reserved.set(usage.reserved.int64)
+        return usage.some
+  )
+
+proc updateBlockMetadata*(
+  self: RepoStore,
+  cid: Cid,
+  plusRefCount: Natural = 0,
+  minusRefCount: Natural = 0,
+  minExpiry: SecondsSince1970 = 0
+): Future[?!void] {.async.} =
+  if cid.isEmpty:
+    return success()
+
+  without metaKey =? createBlockExpirationMetadataKey(cid), err:
+    return failure(err)
+
+  await self.metaDs.modify(metaKey,
+    proc (maybeCurrBlockMd: ?BlockMetadata): Future[?BlockMetadata] {.async.} =
+      if currBlockMd =? maybeCurrBlockMd:
+        BlockMetadata(
+          size: currBlockMd.size,
+          expiry: max(currBlockMd.expiry, minExpiry),
+          refCount: currBlockMd.refCount + plusRefCount - minusRefCount
+        ).some
+      else:
+        raise newException(BlockNotFoundError, "Metadata for block with cid " & $cid & " not found")
+  )
+
+proc storeBlock*(self: RepoStore, blk: Block, minExpiry: SecondsSince1970): Future[?!StoreResult] {.async.} =
+  if blk.isEmpty:
+    return success(StoreResult(kind: AlreadyInStore))
+
+  without metaKey =? createBlockExpirationMetadataKey(blk.cid), err:
+    return failure(err)
+
+  without blkKey =? makePrefixKey(self.postFixLen, blk.cid), err:
+    return failure(err)
+
+  await self.metaDs.modifyGet(metaKey,
+    proc (maybeCurrMd: ?BlockMetadata): Future[(?BlockMetadata, StoreResult)] {.async.} =
+      var
+        md: BlockMetadata
+        res: StoreResult
+
+      if currMd =? maybeCurrMd:
+        if currMd.size == blk.data.len.NBytes:
+          md = BlockMetadata(size: currMd.size, expiry: max(currMd.expiry, minExpiry), refCount: currMd.refCount)
+          res = StoreResult(kind: AlreadyInStore)
+
+          # making sure that the block acutally is stored in the repoDs
+          without hasBlock =? await self.repoDs.has(blkKey), err:
```

Comment:
is this necessary, seems like `modifyGet` would be broken if that wasn't the case?

Reply:
Ah, nevermind, this are different DS.

Reply:
Yes, that's correct.